### DISCUSSION TITLE
Help inserting MDX into JSX

### DISCUSSION THREAD

**Discussion by warrenalphonso** [mdx-js/mdx#1339] [Q&A] [answered]
I'm trying to insert content from `.mdx` files into a `.jsx` file, and I can't figure out how to do this. I'm using Next.js, and auto-generated `.mdx` files work great. Do I have to use `mdx.macro` like in the [CRA setup](https://mdxjs.com/getting-started/create-react-app)? Sorry, it's a really basic question but I haven't gotten anywhere for a few hours.

**Comment by ChristianMurphy** [comment]
Have you tried the MDX+Next guide? https://mdxjs.com/getting-started/next
And/or taken a look at the Next example https://github.com/mdx-js/mdx/tree/master/examples/next
Those should point you in the right direction.

If the issue is persistent even after trying those, could you share a https://codesandbox.io with an example of what you tried?

**Comment by warrenalphonso** [comment, marked as answer]
Thanks for the quick reply! Here's a [git repo](https://github.com/warrenalphonso/mdx-inject) with what I'm trying to do (see `pages/problem.js`). I can reproduce it on a codesandbox if you'd like, I just don't have much experience on codesandbox so went with Git since it'd be faster for me. 

You can run with `yarn dev`, but here's a screenshot of what I've got: 

<img width="1904" alt="Screen Shot 2020-11-16 at 1 17 48 PM" src="https://user-images.githubusercontent.com/38018861/99309743-c255fb00-280e-11eb-9bfa-286feb893d8d.png">

Behavior like what React-Markdown provides is what I'm looking for; it's relatively painless to do `<ReactMarkdown escapeHtml={false} source={file}/>`. I'm able to generate the JSX using `mdx.sync()` but now how do I insert that into the file?

**Reply by ChristianMurphy** [reply]
@warrenalphonso could you expand on why you want to use Next without using the MDX Next loader?

---

If dynamically adding MDX, there is the runtime https://mdxjs.com/advanced/runtime
which works more like react-markdown.

For example https://codesandbox.io/s/mdx-runtime-op6hq

Though be aware, the runtime
> :warning: this is not the preferred way to use MDX since it introduces a substantial amount of overhead and dramatically increases bundle sizes. It should also not be used with user input that isn’t sandboxed.

When possible using a loader like the webpack or next loader would be faster, but it depends on your use case.

**Reply by ChristianMurphy** [reply]
The utility you are using in your example

```js
const mdx = require('@mdx-js/mdx')
```

is a lower level utility.
converts MDX into a JSX document, it is used to create tools like the webpack loader and next loader.

**Reply by warrenalphonso** [reply]
> could you expand on why you want to use Next without using the MDX Next loader?

No idea, I didn’t know I wasnt using it. Can you explain or point me to some docs explaining what the MDX Next loader does? I might be mistaken but I don’t think the articles you linked earlier had anything about this. All I could find was [this](https://nextjs.org/docs/api-reference/next.config.js/custom-webpack-config) but I don’t understand *how* to use it after configuring. 

Happy to try it if it’s better than MDX-runtime.

**Reply by ChristianMurphy** [reply]
It treats mdx documents as next js pages, for example https://github.com/mdx-js/mdx/tree/master/examples/next

**Reply by ChristianMurphy** [reply]
There's the webpack loader https://mdxjs.com/getting-started/webpack which focuses on a general `import Component from 'document.mdx'`
And there's runtime, which loads documents and processes them on the fly in the browser https://mdxjs.com/advanced/runtime

All three have their uses.
It really depends on what you want to do

**Reply by warrenalphonso** [reply]
Oh! I just found out Nextjs has a dynamic import feature, which works perfectly here since MDX supports imports like you pointed out. Thanks!
